InfiniTAM/infinitam_ros_node.cpp

```diff
@@ -190,113 +233,286 @@ bool InfinitamNode::startInfinitam(std_srvs::SetBool::Request& request,
+
+  // write vertices
+  std::size_t v_idx = 0;
+
+  ROS_DEBUG_STREAM("going to fill the mesh with faces.");
+
+  mesh_ptr_->polygons.resize(nr_triangles);
+
+  for (uint i = 0; i < nr_triangles; ++i) {
+    //  Write faces.
+    mesh_ptr_->polygons[i].vertices.resize(3);
+    // The vertex index starts with 1 not with 0 (OBJ-file standard).
+    // the Obj_io.h defines that the vertices start with 0, when writing a file,
+    // it just adds one to each value, so for publishing:
+    // polygon_mesh_ptr->polygons[i].vertices.push_back(i * 3 + 2 + 1);
+    // polygon_mesh_ptr->polygons[i].vertices.push_back(i * 3 + 1 + 1);
+    // polygon_mesh_ptr->polygons[i].vertices.push_back(i * 3 + 0 + 1);
+    // for writing to a file using the library.
+    polygon_mesh_ptr->polygons[i].vertices[0] = (i * 3 + 2);
+    polygon_mesh_ptr->polygons[i].vertices[1] = (i * 3 + 1);
+    polygon_mesh_ptr->polygons[i].vertices[2] = (i * 3 + 0);
+  }
+
+  ROS_INFO_STREAM("cloud filled: header: "
+                  << mesh_ptr_->cloud.header
+                  << "height: " << mesh_ptr_->cloud.height
+                  << " width: " << mesh_ptr_->cloud.width
+                  << " fields.size: " << mesh_ptr_->cloud.fields.size());
+
+  ROS_INFO_STREAM("Polygons vector size: " << mesh_ptr_->polygons.size());
+}
+
+void InfinitamNode::extractITMMeshToPclCloud(
+    const ITMMesh::Triangle* triangleArray,
+    pcl::PointCloud<pcl::PointXYZ>::Ptr point_cloud_pcl) {
+  LOG(INFO) << "extractITMMeshToPclCloud start. ";
+  CHECK_NOTNULL(main_engine_);
+  CHECK_NOTNULL(&point_cloud_pcl);
+
   // Point cloud has at least 3 points per triangle.
   point_cloud_pcl->width = main_engine_->GetMesh()->noTotalTriangles * 3;
   point_cloud_pcl->height = 1;
   point_cloud_pcl->is_dense = false;
-  point_cloud_pcl->points.resize(
-      point_cloud_pcl->width * point_cloud_pcl->height);
+  point_cloud_pcl->points.resize(point_cloud_pcl->width *
+                                 point_cloud_pcl->height);
 
   ROS_ERROR_COND(main_engine_->GetMesh()->noTotalTriangles < 1,
                  "The mesh has too few triangles, only: %d",
                  main_engine_->GetMesh()->noTotalTriangles);
 
+  std::size_t point_number = 0;
+
   // All vertices of the mesh are stored in the pcl point cloud.
   for (int64 i = 0; i < main_engine_->GetMesh()->noTotalTriangles * 3;
-      i = i + 3) {
-    point_cloud_pcl->points[i].x = triangleArray[i].p0.x;
-    point_cloud_pcl->points[i].y = triangleArray[i].p0.y;
-    point_cloud_pcl->points[i].z = triangleArray[i].p0.z;
-
-    point_cloud_pcl->points[i + 1].x = triangleArray[i].p1.x;
-    point_cloud_pcl->points[i + 1].y = triangleArray[i].p1.y;
-    point_cloud_pcl->points[i + 1].z = triangleArray[i].p1.z;
-
-    point_cloud_pcl->points[i + 2].x = triangleArray[i].p2.x;
-    point_cloud_pcl->points[i + 2].y = triangleArray[i].p2.y;
-    point_cloud_pcl->points[i + 2].z = triangleArray[i].p2.z;
+       i = i + 3) {
```

Comment:
Is this iteration correct? It is not the same as used in extractITMMeshToPolygonMesh to get the pcl point cloud from the triangleArray. Anyway, I would use this method (extractITMMeshToPclCloud) to get the pcl cloud in extractITMMeshToPolygonMesh.